subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1282 @@
+			> len - (sizeof(struct ndp16_t) + 2*sizeof(struct ndp16_datagram_t))) {
+		LOG_ERR("  ill position of first ndp: %d (%d)",
+				sys_le16_to_cpu(nth16->wNdpIndex), len);
+		return false;
+	}
+
+	if (sys_le16_to_cpu(nth16->wSequence) != 0
+	    &&  sys_le16_to_cpu(nth16->wSequence) != data->rx_sequence + 1) {
+		LOG_ERR("problem with sequence: %d %d", data->rx_sequence,
+				sys_le16_to_cpu(nth16->wSequence));
+	}
+	data->rx_sequence = sys_le16_to_cpu(nth16->wSequence);
+
+	/*
+	 * check (first) NDP(16)
+	 */
+	const struct ndp16_t *ndp16 = (const struct ndp16_t *)
+			(ntb->data + sys_le16_to_cpu(nth16->wNdpIndex));
+
+	if (sys_le16_to_cpu(ndp16->wLength) < sizeof(struct ndp16_t)
+			+ 2*sizeof(struct ndp16_datagram_t)) {
+		LOG_ERR("  ill ndp16 length: %d", sys_le16_to_cpu(ndp16->wLength));
+		return false;
+	}
+	if (sys_le32_to_cpu(ndp16->dwSignature) != NDP16_SIGNATURE_NCM0
+			&&  sys_le32_to_cpu(ndp16->dwSignature) != NDP16_SIGNATURE_NCM1) {
+		LOG_ERR("  ill signature: 0x%08x",
+				(unsigned int)sys_le32_to_cpu(ndp16->dwSignature));
+		return false;
+	}
+	if (sys_le16_to_cpu(ndp16->wNextNdpIndex) != 0) {
+		LOG_ERR("  cannot handle wNextNdpIndex!=0 (%d)",
+				sys_le16_to_cpu(ndp16->wNextNdpIndex));
+		return false;
+	}
+
+	const struct ndp16_datagram_t *ndp16_datagram = (const struct ndp16_datagram_t *)
+			(ntb->data + sys_le16_to_cpu(nth16->wNdpIndex) + sizeof(struct ndp16_t));
+	int ndx = 0;
+	uint16_t max_ndx = (uint16_t)((sys_le16_to_cpu(ndp16->wLength)
+			- sizeof(struct ndp16_t)) / sizeof(struct ndp16_datagram_t));
+
+	if (max_ndx > CFG_CDC_NCM_RCV_MAX_DATAGRAMS_PER_NTB + 1) {
+		/* number of datagrams in NTB > 1 */
+		LOG_ERR("<<xyx %d (%d)", max_ndx - 1, sys_le16_to_cpu(ntb->nth.wBlockLength));
+	}
+	if (sys_le16_to_cpu(ndp16_datagram[max_ndx-1].wDatagramIndex) != 0
+			||  sys_le16_to_cpu(ndp16_datagram[max_ndx-1].wDatagramLength) != 0) {
+		LOG_DBG("  max_ndx != 0");
+		return false;
+	}
+	while (sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex) != 0
+			&&  sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramLength) != 0) {
+		LOG_DBG("  << %d %d", sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex),
+				sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramLength));
+		if (sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex) > len) {
+			LOG_ERR("(EE) ill start of datagram[%d]: %d (%d)",
+					ndx,
+					sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex), len);
+			return false;
+		}
+		if (sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex)
+				+ sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramLength) > len) {
+			LOG_ERR("(EE) ill end of datagram[%d]: %d (%d)", ndx,
+				sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramIndex)
+				+ sys_le16_to_cpu(ndp16_datagram[ndx].wDatagramLength), len);
+			return false;
+		}
+		++ndx;
+	}
+
+	LOG_HEXDUMP_DBG(ntb->data, len, "NTB");
+
+	/* -> ntb contains a valid packet structure */
+	return true;
+}   /* _cdc_ncm_frame_ok */
+
+
+
+static int cdc_ncm_acl_out_cb(struct usbd_class_data *const c_data,
```

Comment:
This name looks much like TinyUSB, which is a great project! But is misleading in this context, as TinyUSB "callbacks" are defined at compile time, which Zephyr does not support. This suggests that this is to be passed as a function pointer, but it is not.

Reply:
No, this is directly from the ECM driver.  I have used the ECM driver (same folder) as a template and just filled in the relevant NCM stuff.

Reply:
Oh right, I need to give a closer look at ECM. My bad!